rasterio/rio/merge.py

```diff
@@ -8,56 +9,134 @@
+    and are otherwise taken from the first input file.
     """
     import numpy as np
 
     verbosity = (ctx.obj and ctx.obj.get('verbosity')) or 1
     logger = logging.getLogger('rio')
+
     try:
         with rasterio.drivers(CPL_DEBUG=verbosity>2):
             output = files[-1]
             files = files[:-1]
 
             with rasterio.open(files[0]) as first:
+                first_res = first.res
                 kwargs = first.meta
-                kwargs['transform'] = kwargs.pop('affine')
-                dest = np.empty((first.count,) + first.shape, 
-                    dtype=first.dtypes[0])
+                kwargs.pop('affine')
+                nodataval = first.nodatavals[0]
+                dtype = first.dtypes[0]
 
             if os.path.exists(output):
+                # TODO: prompt user to update existing file (-i option) like:
+                # overwrite b.tif? (y/n [n]) n
+                # not overwritten
                 dst = rasterio.open(output, 'r+')
                 nodataval = dst.nodatavals[0]
+                dtype = dst.dtypes[0]
+                dest = np.zeros((dst.count,) + dst.shape, dtype=dtype)
             else:
+                # Create new output file.
+                # Extent from option or extent of all inputs.
+                if not bounds:
+                    # scan input files.
+                    xs = []
+                    ys = []
+                    for f in files:
+                        with rasterio.open(f) as src:
+                            left, bottom, right, top = src.bounds
+                            xs.extend([left, right])
+                            ys.extend([bottom, top])
+                    bounds = min(xs), min(ys), max(xs), max(ys)
+                output_transform = Affine.translation(bounds[0], bounds[3])
+
+                # Resolution/pixel size.
+                if not res:
+                    res = first_res
+                output_transform *= Affine.scale(res[0], -res[1])
+
+                # Dataset shape.
+                output_width = int(math.ceil((bounds[2]-bounds[0])/res[0]))
+                output_height = int(math.ceil((bounds[3]-bounds[1])/res[1]))
+
                 kwargs['driver'] == driver
-                dst = rasterio.open(output, 'w', **kwargs)
-                nodataval = first.nodatavals[0]
+                kwargs['transform'] = output_transform
+                kwargs['width'] = output_width
+                kwargs['height'] = output_height
 
-            dest.fill(nodataval)
+                dst = rasterio.open(output, 'w', **kwargs)
+                dest = np.zeros((first.count, output_height, output_width),
+                        dtype=dtype)
+
+            if nodata is not None:
+                nodataval = nodata
+
+            if nodataval is not None:
+                # Only fill if the nodataval is within dtype's range.
+                inrange = False
+                try:
```

Comment:
@sgillies Consider using a conditional instead of exception handling:

```
try: 
    if np.dtype(dtype).kind in ('i', 'u'):
        info = np.iinfo(dtype)
        inrange = (info.min <= nodataval <= info.max)
    else:
        info = np.finfo(dtype)
        inrange = (info.min <= nodataval <= info.max)

except:
    pass
```